javalib/src/main/scala/java/util/regex/Matcher.scala

```diff
@@ -275,5 +283,358 @@ object Matcher {
+             after (ex: ?, *?, (?=...), (?!..))
+    * @param nodeType The type of the node (Regexleaf, or ParentNode)
+    */
+  private[regex] case class Node(originalGroup: OriginalRegex, var nodeType: NodeType) {
+    var newGroup: Int = 0
+    var matched: String = ""
+    var start = 0
+
+    def transformGroupNumber(mapping: Map[Int, Int]): this.type = {
+      nodeType match {
+        case RegexLeaf(regex)
+            if regex.length >= 2 && regex(0) == '\\' && regex(1).isDigit =>
+          nodeType = RegexLeaf("\\" + mapping.get(regex.substring(1).toInt).getOrElse(0))
+        case RegexLeaf(regex) =>
+        case ParentNode(children) => children.foreach(_.transformGroupNumber(mapping))
+      }
+      this
+    }
+
+    // Assigns consecutive group numbers starting from newGroupIndex to the nodes in this subtree, in a pre-order walk.
+    // Returns 1 plus the largest assigned group number.
+    def setNewGroup(index: Int): Int = {
+      newGroup = index
+      nodeType.fold(
+          r => index + 1,
+          children => children.foldLeft(index+1) {
+            case (newIndex, child) => child.setNewGroup(newIndex)
+          })
+    }
+
+    // Recursively matched groups from a match of this regular expression
+    // to the field matched of each node.
+    // If a group did not match, stringForGroupNumber can return null
+    // This is consistent with JVM regexs
+    def setMatch(stringForGroupNumber: Int => String): Unit = {
+      this.matched = stringForGroupNumber(newGroup)
+      nodeType.fold(
+          r => (),
+          children => children.foreach(_.setMatch(stringForGroupNumber))
+      )
+    }
+
+    // Propagates the start position of this node to its descendants
+    def propagateStart(): Unit = {
+      nodeType.fold(
+        regex => (),
+        children => children.foldLeft(start) {
+          case (newStart, child) => child.setStartReturnEnd(newStart)
+        }
+      )
+    }
+
+    // Propagates the end position of this node to its descendants
+    def propagateEnd(): Unit = {
+      nodeType.fold(
+        regex => (),
+        children => children.foldRight(end) {
+          case (child, newEnd) => child.setEndReturnStart(newEnd)
+        }
+      )
+    }
+
+    def setEndReturnStart(newEnd: Int): Int = {
+      start = originalGroup match {
+        case OriginallyWrapped.Absolute() => newEnd
+        case _ =>
+          if (matched == null) -1
+          else newEnd - matched.length
+      }
+      propagateStart()
+      start
+    }
+    def setStartReturnEnd(offset: Int): Int = {
+      start = if (matched == null) -1
+              else offset
+
+      originalGroup match {
+        case OriginallyWrapped.Repeater(_) => propagateEnd()
+        case _ =>                          propagateStart()
+      }
+      originalGroup match {
+        case OriginallyWrapped.Absolute() => offset
+        case _ => end
+      }
+    }
```

Comment:
That was the trickiest part I had to solve, when assigning group positions. I could not choose between assigning group numbers from left to right or from right to left, because there both failed in one case each. Normally, both directions give the same result. But there are corner cases.
Consider the following regex matching `abbbbbbc`

    (?=ab*(c))ab

After conversion, this becomes:

    (?=(ab*)(c))(ab)

To recover the position of the group (c), we cannot do anything but compute it from the length of (ab*), that is, propagate the start, compute the length, and return the end, and this, recursively. This is what we need to do in a forward-matching regex.

However, consider the following regex matching `abcbdbe`

    a(b.)*

After conversion, it is transformed to:

    (a)((b.)*)

The semantics of group matching under a star are that the last match is considered
Therefore, we cannot obtain the start position of (b.) from propagating lengths from left to right..
What we first do is to get the start, then the end, of the group `((b.)*)`, and then we propagate this end to the inner group.

Note that when javascript will support back-matching `(?<= )` and `(?<! )` (hopefully one day), we will be able to implement the length transfer using the `setEndReturnStart` method, because we have no clue on where the match started (this is different from the `start` position because it can extend before it)

`Absolute()` means a marker of the type `(?= )` or `(?! )`, they do not count to propagate the length on the right or on the left.

`Repeater()` designates all the repeat-transformer which have more or less the semantics of *. Every group having a repeater uses teh semantics of propagating the end to the start.

Reply:
OK, thanks. Looks like you could copy-paste this as a big comment in the code ;)

Reply:
Pasted in the code as a big comment.